README.md
![Color Helper-Dark](branding/dark.png#gh-dark-mode-only)![Color Helper-Light](branding/light.png#gh-light-mode-only)
# Color Helper

[![Latest Version on Packagist](https://img.shields.io/packagist/v/datomatic/color-helper.svg?style=flat-square)](https://packagist.org/packages/datomatic/color-helper)
[![GitHub Tests Action Status](https://img.shields.io/github/workflow/status/datomatic/color-helper/run-tests?label=tests)](https://github.com/datomatic/color-helper/actions?query=workflow%3Arun-tests+branch%3Amain)
[![GitHub Code Style Action Status](https://img.shields.io/github/workflow/status/datomatic/color-helper/Fix%20PHP%20code%20style%20issues?label=code%20style)](https://github.com/datomatic/color-helper/actions?query=workflow%3A"Fix+PHP+code+style+issues"+branch%3Amain)
[![Total Downloads](https://img.shields.io/packagist/dt/datomatic/color-helper.svg?style=flat-square)](https://packagist.org/packages/datomatic/color-helper)

A PHP toolkit class to manage colors functionalities

## Installation

You can install the package via composer:

```bash
composer require datomatic/color-helper
```

## Usage

```php
use Datomatic\Color\Color;
$color = Color::fromHex('#ffffff');
$color->rgb(); // ['r' => 255, 'g' => 255, 'b' => 255] 
$color->hsl(); // ['h' => 0, 's' => 0, 'l' => 100] 
$color->hsv(); // ['h' => 0, 's' => 0, 'v' => 100] 
$color->cmyk(); // ['c' => 0, 'm' => 0, 'y' => 0, 'k' => 0] 
```

## Testing

```bash
composer test
```

## Changelog

Please see [CHANGELOG](CHANGELOG.md) for more information on what has changed recently.

## Contributing

Please see [CONTRIBUTING](https://github.com/datomatic/.github/blob/main/CONTRIBUTING.md) for details.

## Security Vulnerabilities

Please review [our security policy](../../security/policy) on how to report security vulnerabilities.

## Credits

- [Alberto Peripolli](https://github.com/datomatic)
- [All Contributors](../../contributors)

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.


CHANGELOG.md
# Changelog

All notable changes to `color-helper` will be documented in this file.


tests/ConversionsTest.php
<?php

use Datomatic\Color\Color;
use Datomatic\Color\ColorConversion;

it('can convert rgb to hex with #', function (int $r, int $g, int $b, string $hex) {
    $color = Color::fromRgb($r, $g, $b);
    expect($color)->toBeInstanceOf(Color::class);
    expect($color->hex())->toBe($hex);
})->with([
    [0, 158, 204, '#009ecc'],
    [244, 231, 15, '#f4e70f'],
    [0, 0, 0, '#000000'],
    [255, 255, 255, '#ffffff'],
]);

it('can convert rgb to hex withouth #', function (int $r, int $g, int $b, string $hex) {
    $color = Color::fromRgb($r, $g, $b);
    expect($color->hex(false))->toBe($hex);
    expect(ColorConversion::rgbToHex($r, $g, $b))->toBe($hex);
})->with([
    [0, 158, 204, '009ecc'],
    [244, 231, 15, 'f4e70f'],
    [0, 0, 0, '000000'],
    [255, 255, 255, 'ffffff'],
]);

it('can convert hex to rgb', function (string $hex, int $r, int $g, int $b) {
    $color = Color::fromHex($hex);
    expect($color->rgb())->toBe(['r' => $r, 'g' => $g, 'b' => $b]);
})->with([
    ['009ecc', 0, 158, 204],
    ['#f4e70f', 244, 231, 15],
    ['#000', 0, 0, 0],
    ['FFF', 255, 255, 255],
]);


it('can convert rgb to hsv', function (array $rgb, array $hsv) {
    $color = Color::fromRgb(...$rgb);
    expect($color->hsv())->toBe($hsv);
    expect(ColorConversion::rgbToHsv(...$rgb))->toBe($hsv);
})->with([
    [['r' => 66, 'g' => 135, 'b' => 245], ['h' => 217, 's' => 73, 'v' => 96]],
    [['r' => 143, 'g' => 19, 'b' => 62], ['h' => 339, 's' => 87, 'v' => 56]],
    [['r' => 114, 'g' => 207, 'b' => 131], ['h' => 131, 's' => 45, 'v' => 81]],
    [['r' => 0, 'g' => 0, 'b' => 0], ['h' => 0, 's' => 0, 'v' => 0]],
    [['r' => 255, 'g' => 255, 'b' => 255], ['h' => 0, 's' => 0, 'v' => 100]],
]);

it('can convert hsv to rgb', function (array $hsv, array $rgb) {
    $color = Color::fromHsv(...$hsv);
    expect($color->rgb())->toBe($rgb);
    expect(ColorConversion::hsvToRgb(...$hsv))->toBe($rgb);
})->with([
    [['h' => 217, 's' => 73, 'v' => 96], ['r' => 66, 'g' => 135, 'b' => 245]],
    [['h' => 339, 's' => 87, 'v' => 56], ['r' => 143, 'g' => 19, 'b' => 62]],
    [['h' => 131, 's' => 45, 'v' => 81], ['r' => 114, 'g' => 207, 'b' => 131]],
    [['h' => 0, 's' => 0, 'v' => 0], ['r' => 0, 'g' => 0, 'b' => 0]],
    [['h' => 0, 's' => 0, 'v' => 100], ['r' => 255, 'g' => 255, 'b' => 255]],
]);


it('can convert rgb to hsl', function (array $rgb, array $hsl) {
    $color = Color::fromRgb(...$rgb);
    expect($color->hsl())->toBe($hsl);
    expect(ColorConversion::rgbToHsl(...$rgb))->toBe($hsl);
})->with([
    [['r' => 66, 'g' => 135, 'b' => 245], ['h' => 217, 's' => 90, 'l' => 61]],
    [['r' => 143, 'g' => 19, 'b' => 62], ['h' => 339, 's' => 77, 'l' => 32]],
    [['r' => 114, 'g' => 207, 'b' => 131], ['h' => 131, 's' => 49, 'l' => 63]],
    [['r' => 0, 'g' => 0, 'b' => 0], ['h' => 0, 's' => 0, 'l' => 0]],
    [['r' => 255, 'g' => 255, 'b' => 255], ['h' => 0, 's' => 0, 'l' => 100]],
]);


it('can convert hsl to rgb', function (array $hsl, array $rgb) {
    $color = Color::fromHsl(...$hsl);
    expect($color->rgb())->toBe($rgb);
    expect(ColorConversion::hslToRgb(...$hsl))->toBe($rgb);
})->with([
    [['h' => 217, 's' => 90, 'l' => 61], ['r' => 66, 'g' => 135, 'b' => 245]],
    [['h' => 339, 's' => 77, 'l' => 32], ['r' => 144, 'g' => 19, 'b' => 63]],
    [['h' => 290, 's' => 77, 'l' => 63], ['r' => 209, 'g' => 88, 'b' => 233]],
    [['h' => 131, 's' => 49, 'l' => 63], ['r' => 114, 'g' => 207, 'b' => 131]],
    [['h' => 0, 's' => 0, 'l' => 0], ['r' => 0, 'g' => 0, 'b' => 0]],
    [['h' => 0, 's' => 0, 'l' => 100], ['r' => 255, 'g' => 255, 'b' => 255]],
]);


it('can convert rgb to cmyk', function (array $rgb, array $cmyk) {
    $color = Color::fromRgb(...$rgb);
    expect($color->cmyk())->toBe($cmyk);
    expect(ColorConversion::rgbToCmyk(...$rgb))->toBe($cmyk);
})->with([
    [['r' => 66, 'g' => 135, 'b' => 245], ['c' => 73, 'm' => 45, 'y' => 0, 'k' => 4]],
    [['r' => 143, 'g' => 19, 'b' => 62], ['c' => 0, 'm' => 87, 'y' => 57, 'k' => 44]],
    [['r' => 114, 'g' => 207, 'b' => 131], ['c' => 45, 'm' => 0, 'y' => 37, 'k' => 19]],
    [['r' => 0, 'g' => 0, 'b' => 0], ['c' => 0, 'm' => 0, 'y' => 0, 'k' => 100]],
    [['r' => 255, 'g' => 255, 'b' => 255], ['c' => 0, 'm' => 0, 'y' => 0, 'k' => 0]],
]);


it('can convert cmyk to rgb', function (array $cmyk, array $rgb) {
    $color = Color::fromCmyk(...$cmyk);
    expect($color->rgb())->toBe($rgb);
    expect(ColorConversion::cmykToRgb(...$cmyk))->toBe($rgb);
})->with([
    [['c' => 73, 'm' => 45, 'y' => 0, 'k' => 4], ['r' => 66, 'g' => 135, 'b' => 245]],
    [['c' => 0, 'm' => 87, 'y' => 57, 'k' => 44], ['r' => 143, 'g' => 19, 'b' => 61]],
    [['c' => 45, 'm' => 0, 'y' => 37, 'k' => 19], ['r' => 114, 'g' => 207, 'b' => 130]],
    [['c' => 0, 'm' => 0, 'y' => 0, 'k' => 100], ['r' => 0, 'g' => 0, 'b' => 0]],
    [['c' => 0, 'm' => 0, 'y' => 0, 'k' => 0], ['r' => 255, 'g' => 255, 'b' => 255]],
]);


tests/ConstructorsTest.php
<?php

use Datomatic\Color\Color;
use Datomatic\Color\Exceptions\ColorConstructException;

it('can instantiate from string', function (string $string, string $hex) {
    $enum = Color::fromName($string);
    expect($enum)->toBeInstanceOf(Color::class);
    expect($enum->hex())->toBe($hex);
})->with([
    ['aliceblue', '#f0f8ff'],
    ['antiquewhite', '#faebd7'],
    ['aqua', '#00ffff'],
    ['aquamarine', '#7fffd4'],
    ['azure', '#f0ffff'],
    ['beige', '#f5f5dc'],
    ['bisque', '#ffe4c4'],
    ['black', '#000000'],
    ['blanchedalmond', '#ffebcd'],
    ['blue', '#0000ff'],
    ['blueviolet', '#8a2be2'],
    ['brown', '#a52a2a'],
    ['burlywood', '#deb887'],
    ['cadetblue', '#5f9ea0'],
    ['chartreuse', '#7fff00'],
    ['chocolate', '#d2691e'],
    ['coral', '#ff7f50'],
    ['cornflower', '#6495ed'],
    ['cornflowerblue', '#6495ed'],
    ['cornsilk', '#fff8dc'],
    ['crimson', '#dc143c'],
    ['cyan', '#00ffff'],
    ['darkblue', '#00008b'],
    ['darkcyan', '#008b8b'],
    ['darkgoldenrod', '#b8860b'],
    ['darkgray', '#a9a9a9'],
    ['darkgreen', '#006400'],
    ['darkgrey', '#a9a9a9'],
    ['darkkhaki', '#bdb76b'],
    ['darkmagenta', '#8b008b'],
    ['darkolivegreen', '#556b2f'],
    ['darkorange', '#ff8c00'],
    ['darkorchid', '#9932cc'],
    ['darkred', '#8b0000'],
    ['darksalmon', '#e9967a'],
    ['darkseagreen', '#8fbc8f'],
    ['darkslateblue', '#483d8b'],
    ['darkslategray', '#2f4f4f'],
    ['darkslategrey', '#2f4f4f'],
    ['darkturquoise', '#00ced1'],
    ['darkviolet', '#9400d3'],
    ['deeppink', '#ff1493'],
    ['deepskyblue', '#00bfff'],
    ['dimgray', '#696969'],
    ['dimgrey', '#696969'],
    ['dodgerblue', '#1e90ff'],
    ['firebrick', '#b22222'],
    ['floralwhite', '#fffaf0'],
    ['forestgreen', '#228b22'],
    ['fuchsia', '#ff00ff'],
    ['gainsboro', '#dcdcdc'],
    ['ghostwhite', '#f8f8ff'],
    ['gold', '#ffd700'],
    ['goldenrod', '#daa520'],
    ['gray', '#808080'],
    ['green', '#008000'],
    ['greenyellow', '#adff2f'],
    ['grey', '#808080'],
    ['honeydew', '#f0fff0'],
    ['hotpink', '#ff69b4'],
    ['indianred', '#cd5c5c'],
    ['indigo', '#4b0082'],
    ['ivory', '#fffff0'],
    ['khaki', '#f0e68c'],
    ['laserlemon', '#ffff54'],
    ['lavender', '#e6e6fa'],
    ['lavenderblush', '#fff0f5'],
    ['lawngreen', '#7cfc00'],
    ['lemonchiffon', '#fffacd'],
    ['lightblue', '#add8e6'],
    ['lightcoral', '#f08080'],
    ['lightcyan', '#e0ffff'],
    ['lightgoldenrod', '#fafad2'],
    ['lightgoldenrodyellow', '#fafad2'],
    ['lightgray', '#d3d3d3'],
    ['lightgreen', '#90ee90'],
    ['lightgrey', '#d3d3d3'],
    ['lightpink', '#ffb6c1'],
    ['lightsalmon', '#ffa07a'],
    ['lightseagreen', '#20b2aa'],
    ['lightskyblue', '#87cefa'],
    ['lightslategray', '#778899'],
    ['lightslategrey', '#778899'],
    ['lightsteelblue', '#b0c4de'],
    ['lightyellow', '#ffffe0'],
    ['lime', '#00ff00'],
    ['limegreen', '#32cd32'],
    ['linen', '#faf0e6'],
    ['magenta', '#ff00ff'],
    ['maroon', '#800000'],
    ['maroon2', '#7f0000'],
    ['maroon3', '#b03060'],
    ['mediumaquamarine', '#66cdaa'],
    ['mediumblue', '#0000cd'],
    ['mediumorchid', '#ba55d3'],
    ['mediumpurple', '#9370db'],
    ['mediumseagreen', '#3cb371'],
    ['mediumslateblue', '#7b68ee'],
    ['mediumspringgreen', '#00fa9a'],
    ['mediumturquoise', '#48d1cc'],
    ['mediumvioletred', '#c71585'],
    ['midnightblue', '#191970'],
    ['mintcream', '#f5fffa'],
    ['mistyrose', '#ffe4e1'],
    ['moccasin', '#ffe4b5'],
    ['navajowhite', '#ffdead'],
    ['navy', '#000080'],
    ['oldlace', '#fdf5e6'],
    ['olive', '#808000'],
    ['olivedrab', '#6b8e23'],
    ['orange', '#ffa500'],
    ['orangered', '#ff4500'],
    ['orchid', '#da70d6'],
    ['palegoldenrod', '#eee8aa'],
    ['palegreen', '#98fb98'],
    ['paleturquoise', '#afeeee'],
    ['palevioletred', '#db7093'],
    ['papayawhip', '#ffefd5'],
    ['peachpuff', '#ffdab9'],
    ['peru', '#cd853f'],
    ['pink', '#ffc0cb'],
    ['plum', '#dda0dd'],
    ['powderblue', '#b0e0e6'],
    ['purple', '#800080'],
    ['purple2', '#7f007f'],
    ['purple3', '#a020f0'],
    ['rebeccapurple', '#663399'],
    ['red', '#ff0000'],
    ['rosybrown', '#bc8f8f'],
    ['royalblue', '#4169e1'],
    ['saddlebrown', '#8b4513'],
    ['salmon', '#fa8072'],
    ['sandybrown', '#f4a460'],
    ['seagreen', '#2e8b57'],
    ['seashell', '#fff5ee'],
    ['sienna', '#a0522d'],
    ['silver', '#c0c0c0'],
    ['skyblue', '#87ceeb'],
    ['slateblue', '#6a5acd'],
    ['slategray', '#708090'],
    ['slategrey', '#708090'],
    ['snow', '#fffafa'],
    ['springgreen', '#00ff7f'],
    ['steelblue', '#4682b4'],
    ['tan', '#d2b48c'],
    ['teal', '#008080'],
    ['thistle', '#d8bfd8'],
    ['tomato', '#ff6347'],
    ['turquoise', '#40e0d0'],
    ['violet', '#ee82ee'],
    ['wheat', '#f5deb3'],
    ['white', '#ffffff'],
    ['whitesmoke', '#f5f5f5'],
    ['yellow', '#ffff00'],
    ['yellowgreen', '#9acd32'],
]);

it('can instantiate from hex', function (string $hex, string $res) {
    $enum = Color::fromHex($hex);
    expect($enum)->toBeInstanceOf(Color::class);
    expect($enum->hex())->toBe($res);
    expect((string) $enum)->toBe($res);
})->with([
    ['#f0f8ff', '#f0f8ff'],
    ['f0f8ff', '#f0f8ff'],
    ['#ff8', '#ffff88'],
    ['ff8', '#ffff88'],
]);

it('can instantiate from rgb', function (int $r, int $g, int $b, string $res) {
    $enum = Color::fromRgb($r, $g, $b);
    expect($enum)->toBeInstanceOf(Color::class);
    expect($enum->hex())->toBe($res);
})->with([
    [240, 248, 255, '#f0f8ff'],
    [255, 255, 136, '#ffff88'],
]);

it('can instantiate from hsv', function (int $h, int $s, int $v, string $res) {
    $enum = Color::fromHsv($h, $s, $v);
    expect($enum)->toBeInstanceOf(Color::class);
    expect($enum->hex())->toBe($res);
})->with([
    [208, 6, 100, '#f0f8ff'],
    [60, 47, 100, '#ffff87'],
]);

it('can instantiate from hsl', function (int $h, int $s, int $l, string $res) {
    $enum = Color::fromHsl($h, $s, $l);
    expect($enum)->toBeInstanceOf(Color::class);
    expect($enum->hex())->toBe($res);
})->with([
    [208, 100, 97, '#f0f8ff'],
    [60, 100, 76, '#ffff85'],
]);

it('can instantiate from cmyk', function (int $c, int $m, int $y, int $k, string $res) {
    $enum = Color::fromCmyk($c, $m, $y, $k);
    expect($enum)->toBeInstanceOf(Color::class);
    expect($enum->hex())->toBe($res);
})->with([
    [6, 3, 0, 0, '#f0f7ff'],
    [0, 1, 48, 0, '#fffc85'],
]);

it('throw invalidColorName with fromName wrong param', function () {
    expect(fn() => Color::fromName('bob'))
        ->toThrow(ColorConstructException::class, 'Color name `bob` was not recognized');
});

it('throw invalidHexFormat with fromHex wrong param', function () {
    expect(fn() => Color::fromHex('%&GHSN'))
        ->toThrow(ColorConstructException::class, 'HEX color `%&ghsn` doesn\'t match the correct format');
});

it('throw invalidHexLenght with fromHex wrong param', function () {
    expect(fn() => Color::fromHex('ffff'))
        ->toThrow(ColorConstructException::class, 'HEX color `ffff` needs to be 6 or 3 digits long');
});


it('throw invalidIntegerValue with fromRgb wrong param', function (string $param, int $r, int $g, int $b) {
    expect(fn() => Color::fromRgb($r, $g, $b))
        ->toThrow(ColorConstructException::class, "Input param `{$param}` must be between `0` and `255`");
})->with([
    ['r', 300, 200, 10],
    ['g', 34, -4, 300],
    ['b', 100, 4, 280]
]);

it('throw invalidIntegerValue with fromHsv wrong param', function (string $param, int $h, int $s, int $v) {
    expect(fn() => Color::fromHsv($h, $s, $v))
        ->toThrow(ColorConstructException::class, "Input param `{$param}` must be between `0` and `100`");
})->with([
    ['s', 100, 200, 70],
    ['v', 100, 100, 300],
    ['v', 100, 100, -300]
]);


it('not throw invalidIntegerValue with fromHsv h wrong param', function (int $h, int $s, int $v) {
    expect(fn($h, $s, $v) => Color::fromHsv($h, $s, $v))
        ->not->toThrow(ColorConstructException::class);
})->with([
    [-20, 100, 0],
    [380, 0, 100],
]);


it('throw invalidIntegerValue with fromHsl wrong param', function (string $param, int $h, int $s, int $l) {
    expect(fn() => Color::fromHsl($h, $s, $l))
        ->toThrow(ColorConstructException::class, "Input param `{$param}` must be between `0` and `100`");
})->with([
    ['s', 100, 200, 70],
    ['l', 100, 100, 300],
    ['l', 100, 100, -300]
]);


it('not throw invalidIntegerValue with fromHsl h wrong param', function (int $h, int $s, int $l) {
    expect(fn($h, $s, $l) => Color::fromHsl($h, $s, $l))
        ->not->toThrow(ColorConstructException::class);
})->with([
    [-20, 100, 0],
    [380, 0, 100],
]);

it('throw invalidIntegerValue with fromCmyk wrong param', function (string $param, int $c, int $m, int $y, int $k) {
    expect(fn() => Color::fromCmyk($c, $m, $y, $k))
        ->toThrow(ColorConstructException::class, "Input param `{$param}` must be between `0` and `100`");
})->with([
    ['c', -300, 200, 70, 60],
    ['m', 100, 200, 800, 50],
    ['y', 100, 100, -300, 50],
    ['k', 100, 100, 80, 150],
]);


src/ColorConversion.php
<?php

namespace Datomatic\Color;


class ColorConversion
{

    /**
     * @param int $r
     * @param int $g
     * @param int $b
     * @return string
     */
    public static function rgbToHex(int $r, int $g, int $b): string
    {
        return sprintf('%02x%02x%02x', $r, $g, $b);
    }

    /**
     * @param string $hex
     * @return array<string,int>
     */
    public static function hexToRgb(string $hex): array
    {
        $r = hexdec(substr($hex, 0, 2));
        $g = hexdec(substr($hex, 2, 2));
        $b = hexdec(substr($hex, 4, 2));

        return ['r' => (int)round($r), 'g' => (int)round($g), 'b' => (int)round($b)];
    }

    /**
     * @param int $r
     * @param int $b
     * @param int $g
     * @return array<string,int>
     */
    public static function rgbToHsv(int $r, int $b, int $g): array
    {
        $r /= 255;
        $g /= 255;
        $b /= 255;

        $max = max($r, $g, $b);
        $min = min($r, $g, $b);
        $delta = $max - $min;

        $v = 100 * $max;

        if ($delta == 0) {
            $h = 0;
            $s = 0;
        } else {
            $s = 100 * $delta / $max;

            $h = 60 * match ($min) {
                    $r => 3 - (($g - $b) / $delta),
                    $b => 1 - (($r - $g) / $delta),
                    default => 5 - (($b - $r) / $delta),
                };
        }


        return ['h' => (int)round($h), 's' => (int)round($s), 'v' => (int)round($v)];
    }

    /**
     * @param int $h
     * @param int $s
     * @param int $v
     * @return array<string,int>
     */
    public static function hsvToRgb(int $h, int $s, int $v): array
    {
        $rgb = [0, 0, 0];

        for ($i = 0; $i < 4; $i++) {
            if (abs($h - $i * 120) < 120) {
                $distance = max(60, abs($h - $i * 120));
                $rgb[$i % 3] = 1 - (($distance - 60) / 60);
            }
        }

        $max = max($rgb);
        $factor = 255 * ($v / 100);
        for ($i = 0; $i < 3; $i++) {
            $rgb[$i] = (int) round(($rgb[$i] + ($max - $rgb[$i]) * (1 - $s / 100)) * $factor);
        }

        return ['r' => (int)$rgb[0], 'g' => (int)$rgb[1], 'b' => (int)$rgb[2]];
    }

    /**
     * @param int $r
     * @param int $g
     * @param int $b
     * @return array<string,int>
     */
    public static function rgbToHsl(int $r, int $g, int $b): array
    {
        $r /= 255;
        $g /= 255;
        $b /= 255;

        $max = max($r, $g, $b);
        $min = min($r, $g, $b);

        $l = ($max + $min) / 2;
        $d = $max - $min;

        if ($d == 0) {
            $h = $s = 0;
        } else {
            $s = $d / (1 - abs(2 * $l - 1));
            $h = 0;
            switch ($max) {
                case $r:
                    $h = 60 * fmod((($g - $b) / $d), 6);
                    if ($b > $g) {
                        $h += 360;
                    }
                    break;

                case $g:
                    $h = 60 * (($b - $r) / $d + 2);
                    break;

                case $b:
                    $h = 60 * (($r - $g) / $d + 4);
                    break;
            }
        }

        return ['h' => (int)round($h), 's' => (int)round($s * 100), 'l' => (int)round($l * 100)];
    }

    /**
     * @param int $h
     * @param int $s
     * @param int $l
     * @return array<string,int>
     */
    public static function hslToRgb(int $h, int $s, int $l): array
    {
        $h /= 60;
        $h = fmod($h, 6);

        $s = max(0, min(1, $s / 100));
        $l = max(0, min(1, $l / 100));

        $c = (1 - abs((2 * $l) - 1)) * $s;
        $x = $c * (1 - abs(fmod($h, 2) - 1));

        if ($h < 1) {
            $r = $c;
            $g = $x;
            $b = 0;
        } elseif ($h < 2) {
            $r = $x;
            $g = $c;
            $b = 0;
        } elseif ($h < 3) {
            $r = 0;
            $g = $c;
            $b = $x;
        } elseif ($h < 4) {
            $r = 0;
            $g = $x;
            $b = $c;
        } elseif ($h < 5) {
            $r = $x;
            $g = 0;
            $b = $c;
        } else {
            $r = $c;
            $g = 0;
            $b = $x;
        }

        $m = $l - $c / 2;
        $r = (int)round(($r + $m) * 255);
        $g = (int)round(($g + $m) * 255);
        $b = (int)round(($b + $m) * 255);

        return ['r' => $r, 'g' => $g, 'b' => $b];
    }

    /**
     * @param $r
     * @param $g
     * @param $b
     * @return array<string,int>
     */
    public static function rgbToCmyk(int $r, int $g, int $b): array
    {
        $c = 255 - $r;
        $m = 255 - $g;
        $y = 255 - $b;

        $k = min($c, $m, $y);
        if ($k === 255) {
            $c = $m = $y = 0;
        } else {
            $c = ($c - $k) / (255 - $k) * 255;
            $m = ($m - $k) / (255 - $k) * 255;
            $y = ($y - $k) / (255 - $k) * 255;
        }

        return [
            'c' => (int)round($c / 255 * 100),
            'm' => (int)round($m / 255 * 100),
            'y' => (int)round($y / 255 * 100),
            'k' => (int)round($k / 255 * 100)];
    }

    /**
     * @param int $c
     * @param int $m
     * @param int $y
     * @param int $k
     * @return array<string,int>
     */
    public static function cmykToRgb(int $c, int $m, int $y, int $k): array
    {
        $c /= 100;
        $m /= 100;
        $y /= 100;
        $k /= 100;

        $r = 1 - ($c * (1 - $k)) - $k;
        $g = 1 - ($m * (1 - $k)) - $k;
        $b = 1 - ($y * (1 - $k)) - $k;

        $r = (int)round($r * 255);
        $g = (int)round($g * 255);
        $b = (int)round($b * 255);

        return ['r' => $r, 'g' => $g, 'b' => $b];
    }
}


src/ColorNames.php
<?php

namespace Datomatic\Color;

class ColorNames
{
    public const COLORS = [
            'aliceblue' => 'f0f8ff',
            'antiquewhite' => 'faebd7',
            'aqua' => '00ffff',
            'aquamarine' => '7fffd4',
            'azure' => 'f0ffff',
            'beige' => 'f5f5dc',
            'bisque' => 'ffe4c4',
            'black' => '000000',
            'blanchedalmond' => 'ffebcd',
            'blue' => '0000ff',
            'blueviolet' => '8a2be2',
            'brown' => 'a52a2a',
            'burlywood' => 'deb887',
            'cadetblue' => '5f9ea0',
            'chartreuse' => '7fff00',
            'chocolate' => 'd2691e',
            'coral' => 'ff7f50',
            'cornflower' => '6495ed',
            'cornflowerblue' => '6495ed',
            'cornsilk' => 'fff8dc',
            'crimson' => 'dc143c',
            'cyan' => '00ffff',
            'darkblue' => '00008b',
            'darkcyan' => '008b8b',
            'darkgoldenrod' => 'b8860b',
            'darkgray' => 'a9a9a9',
            'darkgreen' => '006400',
            'darkgrey' => 'a9a9a9',
            'darkkhaki' => 'bdb76b',
            'darkmagenta' => '8b008b',
            'darkolivegreen' => '556b2f',
            'darkorange' => 'ff8c00',
            'darkorchid' => '9932cc',
            'darkred' => '8b0000',
            'darksalmon' => 'e9967a',
            'darkseagreen' => '8fbc8f',
            'darkslateblue' => '483d8b',
            'darkslategray' => '2f4f4f',
            'darkslategrey' => '2f4f4f',
            'darkturquoise' => '00ced1',
            'darkviolet' => '9400d3',
            'deeppink' => 'ff1493',
            'deepskyblue' => '00bfff',
            'dimgray' => '696969',
            'dimgrey' => '696969',
            'dodgerblue' => '1e90ff',
            'firebrick' => 'b22222',
            'floralwhite' => 'fffaf0',
            'forestgreen' => '228b22',
            'fuchsia' => 'ff00ff',
            'gainsboro' => 'dcdcdc',
            'ghostwhite' => 'f8f8ff',
            'gold' => 'ffd700',
            'goldenrod' => 'daa520',
            'gray' => '808080',
            'green' => '008000',
            'greenyellow' => 'adff2f',
            'grey' => '808080',
            'honeydew' => 'f0fff0',
            'hotpink' => 'ff69b4',
            'indianred' => 'cd5c5c',
            'indigo' => '4b0082',
            'ivory' => 'fffff0',
            'khaki' => 'f0e68c',
            'laserlemon' => 'ffff54',
            'lavender' => 'e6e6fa',
            'lavenderblush' => 'fff0f5',
            'lawngreen' => '7cfc00',
            'lemonchiffon' => 'fffacd',
            'lightblue' => 'add8e6',
            'lightcoral' => 'f08080',
            'lightcyan' => 'e0ffff',
            'lightgoldenrod' => 'fafad2',
            'lightgoldenrodyellow' => 'fafad2',
            'lightgray' => 'd3d3d3',
            'lightgreen' => '90ee90',
            'lightgrey' => 'd3d3d3',
            'lightpink' => 'ffb6c1',
            'lightsalmon' => 'ffa07a',
            'lightseagreen' => '20b2aa',
            'lightskyblue' => '87cefa',
            'lightslategray' => '778899',
            'lightslategrey' => '778899',
            'lightsteelblue' => 'b0c4de',
            'lightyellow' => 'ffffe0',
            'lime' => '00ff00',
            'limegreen' => '32cd32',
            'linen' => 'faf0e6',
            'magenta' => 'ff00ff',
            'maroon' => '800000',
            'maroon2' => '7f0000',
            'maroon3' => 'b03060',
            'mediumaquamarine' => '66cdaa',
            'mediumblue' => '0000cd',
            'mediumorchid' => 'ba55d3',
            'mediumpurple' => '9370db',
            'mediumseagreen' => '3cb371',
            'mediumslateblue' => '7b68ee',
            'mediumspringgreen' => '00fa9a',
            'mediumturquoise' => '48d1cc',
            'mediumvioletred' => 'c71585',
            'midnightblue' => '191970',
            'mintcream' => 'f5fffa',
            'mistyrose' => 'ffe4e1',
            'moccasin' => 'ffe4b5',
            'navajowhite' => 'ffdead',
            'navy' => '000080',
            'oldlace' => 'fdf5e6',
            'olive' => '808000',
            'olivedrab' => '6b8e23',
            'orange' => 'ffa500',
            'orangered' => 'ff4500',
            'orchid' => 'da70d6',
            'palegoldenrod' => 'eee8aa',
            'palegreen' => '98fb98',
            'paleturquoise' => 'afeeee',
            'palevioletred' => 'db7093',
            'papayawhip' => 'ffefd5',
            'peachpuff' => 'ffdab9',
            'peru' => 'cd853f',
            'pink' => 'ffc0cb',
            'plum' => 'dda0dd',
            'powderblue' => 'b0e0e6',
            'purple' => '800080',
            'purple2' => '7f007f',
            'purple3' => 'a020f0',
            'rebeccapurple' => '663399',
            'red' => 'ff0000',
            'rosybrown' => 'bc8f8f',
            'royalblue' => '4169e1',
            'saddlebrown' => '8b4513',
            'salmon' => 'fa8072',
            'sandybrown' => 'f4a460',
            'seagreen' => '2e8b57',
            'seashell' => 'fff5ee',
            'sienna' => 'a0522d',
            'silver' => 'c0c0c0',
            'skyblue' => '87ceeb',
            'slateblue' => '6a5acd',
            'slategray' => '708090',
            'slategrey' => '708090',
            'snow' => 'fffafa',
            'springgreen' => '00ff7f',
            'steelblue' => '4682b4',
            'tan' => 'd2b48c',
            'teal' => '008080',
            'thistle' => 'd8bfd8',
            'tomato' => 'ff6347',
            'turquoise' => '40e0d0',
            'violet' => 'ee82ee',
            'wheat' => 'f5deb3',
            'white' => 'ffffff',
            'whitesmoke' => 'f5f5f5',
            'yellow' => 'ffff00',
            'yellowgreen' => '9acd32',
        ];
}
